src/compiler/builder.ts

```diff
@@ -0,0 +1,528 @@
+        emitFile(program: Program, path: Path): EmitOutput;
+
+        /** Emit the changed files and clear the cache of the changed files */
+        emitChangedFiles(program: Program): EmitOutputDetailed[];
+        /** When called gets the semantic diagnostics for the program. It also caches the diagnostics and manage them */
+        getSemanticDiagnostics(program: Program, cancellationToken?: CancellationToken): Diagnostic[];
+
+        /** Called to reset the status of the builder */
+        clear(): void;
+    }
+
+    interface EmitHandler {
+        /**
+         * Called when sourceFile is added to the program
+         */
+        onAddSourceFile(program: Program, sourceFile: SourceFile): void;
+        /**
+         * Called when sourceFile is removed from the program
+         */
+        onRemoveSourceFile(path: Path): void;
+        /**
+         * Called when sourceFile is changed
+         */
+        onUpdateSourceFile(program: Program, sourceFile: SourceFile): void;
+        /**
+         * Called when source file has not changed but has some of the resolutions invalidated
+         * If returned true, builder will mark the file as changed (noting that something associated with file has changed)
+         */
+        onUpdateSourceFileWithSameVersion(program: Program, sourceFile: SourceFile): boolean;
+        /**
+         * Gets the files affected by the script info which has updated shape from the known one
+         */
+        getFilesAffectedByUpdatedShape(program: Program, sourceFile: SourceFile, singleFileResult: string[]): string[];
+    }
+
+    export function getFileEmitOutput(program: Program, sourceFile: SourceFile, emitOnlyDtsFiles: boolean, isDetailed: boolean,
+        cancellationToken?: CancellationToken, customTransformers?: CustomTransformers): EmitOutput | EmitOutputDetailed {
+        const outputFiles: OutputFile[] = [];
+        let emittedSourceFiles: SourceFile[];
+        const emitResult = program.emit(sourceFile, writeFile, cancellationToken, emitOnlyDtsFiles, customTransformers);
+        if (!isDetailed) {
+            return { outputFiles, emitSkipped: emitResult.emitSkipped };
+        }
+
+        return {
+            outputFiles,
+            emitSkipped: emitResult.emitSkipped,
+            diagnostics: emitResult.diagnostics,
+            sourceMaps: emitResult.sourceMaps,
+            emittedSourceFiles
+        };
+
+        function writeFile(fileName: string, text: string, writeByteOrderMark: boolean, _onError: (message: string) => void, sourceFiles: SourceFile[]) {
+            outputFiles.push({ name: fileName, writeByteOrderMark, text });
+            if (isDetailed) {
+                emittedSourceFiles = addRange(emittedSourceFiles, sourceFiles);
+            }
+        }
+    }
+
+    interface FileInfo {
+        fileName: string;
+        version: string;
+        signature: string;
+    }
+
+    export function createBuilder(

```

Comment:
Do we want this to be a new public API?
(If so it would be a good idea to take an options object instead of taking many positional arguments.)